mailbox/api/src/test/java/org/apache/james/mailbox/events/EventBusContract.java

```diff
@@ -0,0 +1,263 @@
+
+        eventBus().dispatch(event, ImmutableSet.of()).join();
+        verifyZeroInteractions(listener);
+    }
+
+    @Test
+    default void registerShouldThrowWhenAGroupIsAlreadyUsed() {
+        MailboxListener listener = newListener();
+        MailboxListener listener2 = newListener();
+
+        eventBus().register(listener, new GroupA());
+
+        assertThatThrownBy(() -> eventBus().register(listener2, new GroupA()))
+            .isInstanceOf(GroupAlreadyRegistered.class);
+    }
+
+    @Test
+    default void registerShouldNotThrowOnAnUnregisteredGroup() {
+        MailboxListener listener = newListener();
+        MailboxListener listener2 = newListener();
+
+        eventBus().register(listener, new GroupA()).unregister();
+
+        assertThatCode(() -> eventBus().register(listener2, new GroupA()))
+            .doesNotThrowAnyException();
+    }
+
+    @Test
+    default void dispatchShouldNotImpactListenerRegisteredWhenEmpty() {
+        MailboxListener listener = newListener();
+        eventBus().register(listener, new MailboxIdRegistrationKey(ID_1));
+
+        eventBus().dispatch(event, ImmutableSet.of()).join();
+
+        verifyZeroInteractions(listener);
+    }
+
+    @Test
+    default void dispatchShouldNotImpactListenerRegisteredOnOtherKeys() {
+        MailboxListener listener = newListener();
+        eventBus().register(listener, new MailboxIdRegistrationKey(ID_1));
+
+        eventBus().dispatch(event, ImmutableSet.of(new MailboxIdRegistrationKey(ID_2))).join();
+
+        verifyZeroInteractions(listener);
+    }
+
+    @Test
+    default void dispatchShouldImpactListenerRegistered() {
+        MailboxListener listener = newListener();
+        eventBus().register(listener, new MailboxIdRegistrationKey(ID_1));
+
+        eventBus().dispatch(event, ImmutableSet.of(new MailboxIdRegistrationKey(ID_1))).join();
+
+        verify(listener, times(1)).event(any());
+    }
+
+    @Test
+    default void dispatchShouldImpactOnlyListenerRegistered() {
+        MailboxListener listener = newListener();
+        MailboxListener listener2 = newListener();
+        eventBus().register(listener, new MailboxIdRegistrationKey(ID_1));
+        eventBus().register(listener2, new MailboxIdRegistrationKey(ID_2));
+
+        eventBus().dispatch(event, ImmutableSet.of(new MailboxIdRegistrationKey(ID_1))).join();
+
+        verify(listener, times(1)).event(any());
+        verifyZeroInteractions(listener2);
+    }
+
+    @Test
+    default void registerShouldAllowDuplicatedRegistration() {
+        MailboxListener listener = newListener();
+        eventBus().register(listener, new MailboxIdRegistrationKey(ID_1));
+        eventBus().register(listener, new MailboxIdRegistrationKey(ID_1));
+
+        eventBus().dispatch(event, ImmutableSet.of(new MailboxIdRegistrationKey(ID_1))).join();
+
+        verify(listener, times(1)).event(any());
+    }
+
+    @Test
+    default void unregisterShouldNotRemoveDoubleRegisteredListener() {
```

Comment:
I would rather unregister the listener.
IMO register is just idempotent but doesn't keep track of how many register happened for a given listener